src/index.js

```diff
@@ -47,190 +70,109 @@ class PinoWebpackPlugin {
         }
       })
 
-      // When webpack has finished analyzing and bundling all files, compile marked external files
-      compilation.hooks.processAssets.tapAsync(
-        {
-          name: 'PinoWebpackPlugin',
-          stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL
-        },
-        (assets, callback) => {
-          const childCompiler = this.createChildCompiler(compiler, compilation, workers)
-
-          // When the files have been compiled, add a compatibility footer
-          childCompiler.hooks.thisCompilation.tap('PinoWebpackPlugin', (childCompilation) => {
-            childCompilation.hooks.processAssets.tapAsync(
-              {
-                name: 'PinoWebpackPlugin',
-                stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL
-              },
-              this.addCompatibilityFooter.bind(this)
-            )
-          })
-
-          // Perform the compilation and then track each generated file relative path
-          childCompiler.run((err, stats) => {
-            /* c8 ignore next 3 */
-            if (err) {
-              return callback(err)
+      /** Compile dependencies and add result to compilation assets **/
+      compilation.hooks.additionalAssets.tapAsync(PLUGIN_NAME, (processAssetsCallback) => {
+        // Create child compiler for dependencies
+        const childCompiler = compilation.createChildCompiler(
+          `${PLUGIN_NAME}ChildCompiler`,
+          {
+            library: {
+              type: 'commonjs2'
+            },
+            iife: false
+          },
+          Object.values(dependencies).map((worker) => worker.entry)
+        )
+        childCompiler.inputFileSystem = compiler.inputFileSystem
+        // Generate files in memory
+        // It will be written to output path by main compilation
+        childCompiler.outputFileSystem = fs
```

Comment:
Hey @gdorsi

I thought that avoiding writing intermediate assets to the file system and keeping them in memory would be a good idea...

Actually, not all the running tests are passing: **the cached build test doesn't pass** (this is why I tagged you in my comment). The second build is generating new files...

Noted for changing style in PRs. Actually, it helped me to understand the process of the plugin, so for me it was kinda part of the feature...

Could you take a look on the cache handling part? And if you could explain to me how it works...

Reply:
And the EntryPlugin was already there as I remember (`addExternalFile` and `trackInclusions` functions)

Reply:
Still not for me:
```
test/cached-build.test.js
  it should correctly generate all required pino files when the cache is enabled on Webpack
    ✓ should not error
    ✓ should not error
    ✓ expect falsey value
    ✓ expect truthy value
    1) expect truthy value
    ✓ expect truthy value
    ✓ expect truthy value
    ✓ should match pattern provided
    ✓ should match pattern provided
```

Maybe there's a cache folder I did'nt clear, but I don't see any...

Reply:
> Noted for changing style in PRs. Actually, it helped me to understand the process of the plugin, so for me it was kinda part of the feature...

I also do that when I'm facing new code :smile:, but once you have reached a working solution it is good to go back and minimize the number of changes.

It's just a matter to help each other to get the best possible result, especially for this kind of library that is both complex and rarely updated.

Reply:
This is the file structure when I output the cahed-build test to not temp folder
<img width="336" alt="image" src="https://user-images.githubusercontent.com/15171326/213170617-af6aed8d-b8ea-43bb-8a36-043f44695bcf.png">

Reply:
> > Noted for changing style in PRs. Actually, it helped me to understand the process of the plugin, so for me it was kinda part of the feature...
> 
> I also do that when I'm facing new code 😄, but once you have reached a working solution it is good to go back and minimize the number of changes.
> 
> It's just a matter to help each other to get the best possible result, especially for this kind of library that is both complex and rarely updated.

Yes, I'll know that for my next PRs...
it's my first contribution to a project, thanks for understanding

Reply:
The tests on CI [are passing](https://github.com/pinojs/pino-webpack-plugin/actions/runs/3948483588/jobs/6758620789#step:6:20) so it should be something on your local setup.

Reply:
> I thought that avoiding writing intermediate assets to the file system and keeping them in memory would be a good idea...

This kind of optimization may come already from webpack, in order to avoid issues on edge cases (e.g. the dev server) I suggest reverting to the standard one.

Reply:
That's weird: I tried to run it in a devcontainer to be independant from local system, and the test still doesn't pass...

Reply:
> > I thought that avoiding writing intermediate assets to the file system and keeping them in memory would be a good idea...
> 
> This kind of optimization may come already from webpack, in order to avoid issues on edge cases (e.g. the dev server) I suggest reverting to the standard one.

Actually, we have two files generated if you don't use this: the original one is written to the file system (without the compatibility footer), but not deleted when you create the main compilation asset with the footer added... we would have to handle manually the original file deletion....

Reply:
Actually, it could be a good thing to rewrite the tests not to only check that the expected files are here, but also that ONLY the expected files are here... also by checking that each file is an asset on the resulting webpack stats, etc... 

(and more human friendly description would be nice too 😇)

I'll could make a PR if you think it's a good idea...

Reply:
> Could you take a look on the cache handling part? And if you could explain to me how it works.

When the cache is enabled `trackInclusions` isn't called for the files that are not changed.
So the `handleCache` method kicks in to restore the missing modules from the cache.

> That's weird: I tried to run it in a devcontainer to be independant from local system, and the test still doesn't pass...

So it may be related to the environment or the dependencies.

I'm struggling to understand what you have done, could you either explain your solution or remove the style changes?

Reply:
Found the issue, the tests are passing because of the `memoryFs` that is created at the module level.

Reply:
The main problem in the current code is:
- The child compiler is fully independant: it adds the compatibility footer after emitting its assets to files, then store the generated paths.
- The main compiler add content (banner) to the entrypoints files, but is actually not aware of what files have been generated by the child compiler (except in the banner added)
- Some code could be in a more appropriate hook (`processAssets` with `PROCESS_ASSETS_STAGE_ADDITIONS` stage to add content to an asset for example)

**This leads to the issue #58** : other plugins never know that this plugin created new files as they are not parts of the "assets" property of the compilation stats. And for '"destructive" plugins (like CleanWebpackPlugin), this means the new files are deleted

To be able to avoid that, we have to use a different flow: we cannot add the assets on the child compiler callback as those assets are already emmited, that's why I have to do this in the child compiler's `chunkAsset` and `processAssets`hook.
I can do this again by letting the functions which are outside of the main plugin, but I'd have to change their bodies and arguments, it seem to me that it's gonna be as hard to review it...

As you like, I don't wan't to interfere with the project...

Reply:
Thanks for the explanation :smile: 

Switching back to `childCompiler.outputFileSystem = compiler.outputFileSystem` breaks a few tests on my local machine.

To be sure that this is a real issue, could you try the same on your local?

Reply:
I confirm that it breaks some tests: as there is now several file like "pino-file-xxxx.js", and the test is choosing the first one in alphabetical order...

Reply:
Ah ok, I missed [this](https://github.com/pinojs/pino-webpack-plugin/pull/59#discussion_r1073482267) so now I understand why the memory fs is useful (would be worth a comment in the code).

Good call!

Have you find out why the tests are not passing on your local?

If not I could try to help.

Can you share your node version and the lockfile?

Reply:
The tests are now passing on my local computer **BUT SHOULDN'T**: two 'first-xxxx.js' and 'abc/cde/second-xxxxx.js' files are generated...
As `readdirSync` is returning filenames in alphabetical order, depending on the hash, tests could succeed when it shouldn't... for example they should fail if there's several file starting with 'first-'...

A way to help me would be to test with no temp dir, look at the output and see why the file generated on the second build with cache enabled has a different name: as I said, I don't really understand how the webpack cache works and there's poor documentation on that... does the childCompiler use cache?

Reply:
> Ah ok, I missed [this](https://github.com/pinojs/pino-webpack-plugin/pull/59#discussion_r1073482267) so now I understand why the memory fs is useful (would be worth a comment in the code) and why the tests fail if it is disabled.

You're perfectly right, I added a comment on that in a new commit, let me know if it's clear enough

Reply:
> look at the output and see why the file generated on the second build with cache enabled has a different name

If the name is generated using the content hash it is normal for them to have different names.

>  I don't really understand how the webpack cache works and there's poor documentation on that... does the childCompiler use cache?

The poor docs is what make writing Webpack plugins so hard.
I did this one year ago and don't remember these parts, only that you need to check the Webpack source code to understand what's happening.

Reply:
> > look at the output and see why the file generated on the second build with cache enabled has a different name
> 
> If the name is generated using the content hash it is normal for them to have different names.

What I don't understand is that file content hasn't changed, so the hash shouldn't have either...
I'm gonna try to dive deeper in the cache handling, but don't know when I'll be able to fix this...